src/manifest/container.rs

```diff
@@ -0,0 +1,490 @@
+    ///
+    /// `with_type()` and `with_metadata()` must have been called before this
+    /// function is called; otherwise, an error will be returned.
+    ///
+    /// [`Container`]: struct.Container.html
+    pub fn sign(
+        mut self,
+        signer: &mut impl rsa::Signer,
+    ) -> Result<&'m mut [u8], Error> {
+        if !self.has_type || !self.has_metadata {
+            return Err(Error::OutOfRange);
+        }
+
+        let sig_len = signer.pub_len().byte_len();
+        let total_len = self
+            .cursor
+            .consumed_len()
+            .checked_add(sig_len)
+            .ok_or(io::Error::BufferExhausted)?;
+
+        let u16_max = u16::MAX as usize;
+        if total_len > u16_max || sig_len > u16_max {
+            return Err(Error::OutOfRange);
+        }
+
+        let mark = self.cursor.consumed_len();
+        self.cursor.seek(LEN_OFFSET)?;
+        self.cursor.write_le(total_len as u16)?;
+        self.cursor.seek(SIG_LEN_OFFSET)?;
+        self.cursor.write_le(sig_len as u16)?;
+        // Always set the "padding" bytes to 0xff.
+        self.cursor.write_le(0xffffu16)?;
+        self.cursor.seek(mark)?;
+
+        let (message, sig) = self
+            .cursor
+            .consume_with_prior(sig_len)
+            .ok_or(io::Error::BufferExhausted)?;
+        signer
+            .sign(message, sig)
+            .map_err(|_| Error::SignatureFailure)?;
+        Ok(self.cursor.take_consumed_bytes())
+    }
+}
+
+impl Write for Containerizer<'_> {
+    fn write_bytes(&mut self, buf: &[u8]) -> Result<(), io::Error> {
+        self.cursor.write_bytes(buf)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use static_assertions::const_assert_eq;
+
+    use crate::crypto::ring;
+    use crate::crypto::rsa::Builder as _;
+    use crate::crypto::rsa::Keypair as _;
+    use crate::crypto::rsa::Signer as _;
+    use crate::crypto::rsa::SignerBuilder as _;
+    use crate::crypto::testdata;
+
+    const MANIFEST_HEADER: &[u8] = &[
+        0x1f, 0x01, // Total length. This is the header length (12) +
+        //          // body length (19) + signature length (256).
+        0x0e, 0xda, // FPM magic.
+        0xaa, 0x55, 0x01, 0x00, // Container id (0x155aa).
+        0x00, 0x01, // Signature length (0x100 = 256).
+        0xff, 0xff, // Padding to 4 bytes.
+    ];
+
+    const MANIFEST_CONTENTS: &[u8] = b"Container contents!";
+    const_assert_eq!(MANIFEST_CONTENTS.len(), 19);
+
+    const MANIFEST_LEN: usize =
+        MANIFEST_HEADER.len() + MANIFEST_CONTENTS.len() + 256;
+
+    #[test]
+    fn parse_manifest() {
+        let keypair =
+            ring::rsa::Keypair::from_pkcs8(testdata::RSA_2048_PRIV_PKCS8)
+                .unwrap();
+        let pub_key = keypair.public();
+        let rsa_builder = ring::rsa::Builder::new();
+        let mut signer = rsa_builder.new_signer(keypair).unwrap();
+        let mut rsa = rsa_builder.new_engine(pub_key).unwrap();
```

Comment:
Can you refactor this part (potentially other lines as well) to a utility function in `mod test` to avoid repetition?

Reply:
Yup!